### PR TITLE
Add GitHub Actions CI for Windows

### PR DESCRIPTION
**Pull request by Margen67**

### CODE REVIEW

**Comment by mastercoms** [comment]
Rebase with the latest, 700fdac fixes.

**Comment by Margen67** [comment]
> 
> 
> Rebase with the latest, [700fdac](https://github.com/mastercomfig/team-comtress-2/commit/700fdaca4d0012ecda02bc47f858655f0d276831) fixes.

Still fails:
https://github.com/Margen67/team-comtress-2/runs/1121478014?check_suite_focus=true#step:6:9417
```
"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\game\client\client_tf.vcxproj" (default target) (13) ->
(Link target) -> 
  LINK : fatal error LNK1181: cannot open input file '..\..\lib\common\win32\2015\release\cryptlib.lib' [D:\a\team-comtress-2\team-comtress-2\game\client\client_tf.vcxproj]


"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\dedicated\dedicated.vcxproj" (default target) (26) ->
  LINK : fatal error LNK1181: cannot open input file '..\lib\common\win32\2015\release\cryptlib.lib' [D:\a\team-comtress-2\team-comtress-2\dedicated\dedicated.vcxproj]


"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\engine\engine.vcxproj" (default target) (29) ->
  LINK : fatal error LNK1181: cannot open input file '..\lib\common\win32\2015\release\binkw32.lib' [D:\a\team-comtress-2\team-comtress-2\engine\engine.vcxproj]


"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\filesystem\filesystem_stdio.vcxproj" (default target) (30) ->
  LINK : fatal error LNK1181: cannot open input file '..\lib\common\win32\2015\release\cryptlib.lib' [D:\a\team-comtress-2\team-comtress-2\filesystem\filesystem_stdio.vcxproj]


"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\gameui\GameUI.vcxproj" (default target) (31) ->
  LINK : fatal error LNK1181: cannot open input file '..\lib\common\win32\2015\release\cryptlib.lib' [D:\a\team-comtress-2\team-comtress-2\gameui\GameUI.vcxproj]

"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\replay\replay.vcxproj" (default target) (40) ->
  LINK : fatal error LNK1181: cannot open input file '..\lib\common\win32\2015\release\libjpeg.lib' [D:\a\team-comtress-2\team-comtress-2\replay\replay.vcxproj]


"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\game\server\server_tf.vcxproj" (default target) (42) ->
  LINK : fatal error LNK1181: cannot open input file '..\..\lib\common\win32\2015\release\cryptlib.lib' [D:\a\team-comtress-2\team-comtress-2\game\server\server_tf.vcxproj]


"D:\a\team-comtress-2\team-comtress-2\games.sln" (default target) (1) ->
"D:\a\team-comtress-2\team-comtress-2\materialsystem\shaderapidx9\shaderapidx9.vcxproj" (default target) (44) ->
  LINK : fatal error LNK1181: cannot open input file '..\..\lib\common\win32\2015\release\d3dx9.lib' [D:\a\team-comtress-2\team-comtress-2\materialsystem\shaderapidx9\shaderapidx9.vcxproj]
```

**Comment by mastercoms** [comment]
Ok, should FINALLY be fixed.

**Comment by Margen67** [comment]
> 
> 
> Ok, should FINALLY be fixed. Just run `download_libs.bat` as a step to download some of the libs too large for GitHub.

`download_libs.bat` is broken, but I made a pull request to fix it:  #186
With that change it gets further, but still fails:
https://github.com/Margen67/team-comtress-2/runs/1123575735#step:7:11381
```
cryptlib.lib(randpool.obj) : error LNK2001: unresolved external symbol __CrtDbgReportW [D:\a\team-comtress-2\team-comtress-2\filesystem\filesystem_stdio.vcxproj]
```

**Comment by mastercoms** [comment]
Probably missing MSVC dependencies.

**Comment by mastercoms** [comment]
Sorry, I had uploaded the debug version of cryptlib to the release folder. It should be fixed now.

**Comment by Margen67** [comment]
> 
> 
> Sorry, I had uploaded the debug version of cryptlib to the release folder. It should be fixed now.

Release now builds, but Debug still fails:
https://github.com/Margen67/team-comtress-2/runs/1124900032#step:7:11245
```
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: The command "call ..\vpc_scripts\valve_p4_edit.cmd ..\..\game\.\hl2.exe .. [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: call ..\vpc_scripts\valve_p4_edit.cmd ..\..\game\.\default.pdb .. [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: copy D:\a\team-comtress-2\team-comtress-2\launcher_main\.\Debug\.\default.exe ..\..\game\.\hl2.exe [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: if ERRORLEVEL 1 goto BuildEventFailed [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: copy D:\a\team-comtress-2\team-comtress-2\launcher_main\Debug\\default.pdb ..\..\game\.\default.pdb [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: if ERRORLEVEL 1 goto BuildEventFailed [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: if exist D:\a\team-comtress-2\team-comtress-2\launcher_main\Debug\default.map copy D:\a\team-comtress-2\team-comtress-2\launcher_main\Debug\default.map ..\..\game\.\hl2.map [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: call ..\vpc_scripts\valve_p4_edit.cmd ..\..\game\.\hl2.dat .. [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: ..\..\game\.\bin\newdat ..\..\game\.\hl2.exe [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: goto BuildEventOK [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: :BuildEventFailed [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: echo *** ERROR! PostBuildStep FAILED for launcher_main! EXE or DLL is probably running. *** [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: del /q D:\a\team-comtress-2\team-comtress-2\launcher_main\.\Debug\.\default.exe [D:\a\team-comtress-2\team-comtress-2\launcher_main\launcher_main.vcxproj]
```

**Comment by Margen67** [comment]
Debug still fails after https://github.com/mastercomfig/team-comtress-2/commit/9f24cdbee917634a10e510c3e8bba2799d154c42:
https://github.com/Margen67/team-comtress-2/runs/1136565504?check_suite_focus=true#step:8:11381

**Comment by Margen67** [comment]
Release build is failing now: https://github.com/Margen67/team-comtress-2/runs/1139400712?check_suite_focus=true#step:8:3937

**Comment by maximsmol** [comment]
We require C++17 with the latest threading fixes. VPC files were updated to match, I think

**Comment by Margen67** [comment]
> 
> 
> We require C++17 with the latest threading fixes. VPC files were updated to match, I think

The [build instructions](https://github.com/mastercomfig/team-comtress-2/wiki/Windows-Build-Instructions#building) tell you to `Select all projects. Go to Properties > C/C++ > Language > C++ Language Standard and set language standard to C++17.`

It might be possible to add it to each `vcxproj` manually as a workaround if a fix isn't possible.

**Comment by Margen67** [comment]
I filed new issues related to this:
  * #318
  * #319

**Comment by melvyn2** [comment]
shouldn't the assets from `game_clean` also be added to the artifact zips?

**Comment by Margen67** [comment]
> 
> 
> shouldn't the assets from `game_clean` also be added to the artifact zips?

I added it as a separate artifact.

**Comment by mastercoms** [comment]
https://github.com/mastercomfig/team-comtress-2/pull/165#issuecomment-696522488

Can you push these changes?

**Comment by Margen67** [comment]
> 
> 
> [#165 (comment)](https://github.com/mastercomfig/team-comtress-2/pull/165#issuecomment-696522488)
> 
> Can you push these changes?

That comment was outdated. I've pushed it since then.

**Comment by mastercoms** [comment]
Is it possible to add cache to the built files? I don't want a full 24 minute build on our CI every commit.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          msbuild libprotobuf.vcxproj /p:Configuration=Release
+          msbuild libprotobuf.vcxproj /p:Configuration=Debug
+      - name: Download libs
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: .\download_libs.bat
+      - name: Create game projects
+        run: |
+          reg add "HKLM\SOFTWARE\WOW6432Node\Microsoft\VisualStudio\10.0\Projects\{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}" /v DefaultProjectExtension /t REG_SZ /d vcproj /f
+          ${{ matrix.creategameprojects }}
+          Start-Process "${env:PROGRAMFILES(X86)}\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\devenv.com" "games.sln /upgrade" -nnw -Wait
```
Comment (mastercoms) [inline .github/workflows/CI.yml:54]: What does this upgrade line do? Seems like the CI output says it isn't needed?

**Review by mastercoms — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml (hunk 1 above)
Comment (Margen67) [reply .github/workflows/CI.yml:54]: It makes the sln work with msbuild: #318

**Comment by Margen67** [comment]
> 
> 
> Is it possible to add cache to the built files? I don't want a full 33 minute build on our CI every commit.

Probably, but it depends on what you want cached.

**Comment by mastercoms** [comment]
Ideally, enough for incremental builds/linking. But the goal is to get build times and redundant work down as much as possible.

**Comment by Margen67** [comment]
> 
> 
> Ideally, enough for incremental builds/linking. But the goal is to get build times and redundant work down as much as possible.

I cached libprotobuf, `iobj`, `ipdb`, and the build time went down to ~22m. Is there anything else I should cache?

File: .github/workflows/CI.yml (hunk 1 above)
Comment (mastercoms) [reply .github/workflows/CI.yml:54]: Can we cache vcxproj, vpc and vpc.sentinel then?

Comment (Juesto) [reply .github/workflows/CI.yml]: the vpc crc and sentinel will fail the check; vpc also adds a project dir preprocessor to the projects, you could cache a vpc-less vcxproj and it'll work but manual maintenance will be needed.
I believe reasons like those is why it's a good idea to either have VS2015 installed or add later VS support on VPC natively.

**Comment by mastercoms** [comment]
Try `*.pch` too

File: .github/workflows/CI.yml (hunk 1 above)
Comment (mastercoms) [reply .github/workflows/CI.yml:54]: Depends on when it is cached.

**Comment by Margen67** [comment]
> 
> 
> Try `*.pch`, `*.ipdb` too

https://github.com/Margen67/team-comtress-2/runs/1185896885

File: .github/workflows/CI.yml (hunk 1 above)
Comment (Margen67) [reply .github/workflows/CI.yml:54]: What Momentum Mod does is add to the sln:
https://github.com/momentum-mod/game/blob/develop/mp/src/sln_fix.txt
https://github.com/momentum-mod/game/blob/develop/.github/workflows/CI.yml#L27

This is much faster than upgrading the sln.

**Comment by Margen67** [comment]
This now depends on #431.

**Pull request closed without merging**